DietPi-Software | Paper MC Server

nonnorm opened the pull request:
<!--
Before submitting a pull request:
- Please ensure the target branch is "dev" (active development): https://github.com/MichaIng/DietPi/tree/dev
- Please ensure changes have been tested and verified functional.
-->
**Status**:  Ready
- [x] Figure out how to run server to generate configuration files, then stop.
- [x] Configure using those files
- [x] Figure out why after uninstall + reinstall service not found
- [x] See if stop time can possibly be reduced
(Figure out how to inject stop into the jar?) (MCrcon)
- [x] Test on VM (upgraded to 2GiB of RAM)
- [x] Add to software list on dietpi.com/fix on p#p=

**Reference**: https://github.com/MichaIng/DietPi/issues/3823

**Commit list/description**:
<!--
- DietPi-Config | Add "Fan control" option to "Performance Options"
![Screenshot](https://xxx.github.com/images/xxx.png)
- DietPi-Config | Add "Fan control" support for Odroid C2
- DietPi-Config | Syntax fix
-->
- DietPi-Software | Add Paper Minecraft server to list, install, and configure
- DietPi-Software | Add Paper Minecraft server to uninstall
- DietPi-Software | Fixes
- DietPi-Software | Create config files
- DietPi-Software | Add floodgate authentication
- DietPi-Software | More Fixes
- DietPi-Software | Remove apostrophe
- DietPi-Software | Make use of new function
- DietPi-Software | Fix uninstall
- DietPi-Software | Cosmetic fixes
- DietPi-Software | Reverse CFG command
- DietPi-Software | Increase timeout
- DietPi-Software | Use G_CONFIG_INJECT EULA
- DietPi-Software | Add G_EXEC_OUTPUT
- DietPi-Software | Add G_EXEC_NOEXIT
- DietPi-Software | Add 1 second of sleep
- DietPi-Software | Switch back to echo
- DietPi-Software | Switch back to G_CONFIG_INJECT
- DietPi-Software | Add cd
- DietPi-Software | Add G_EXEC_OUTPUT
- DietPi-Software | Give users access to /opt/papermc
- DietPi-Software | Fix uninstall again
- DietPi-Software | Increase timeout time
- DietPi-Software | Add Config_Create for Geyser CFG
- DietPi-Software | Add missing service
- DietPi-Software | Add blank line
- DietPi-Services | Add to dietpi-services
- DietPi-Software | Raise Geyser timeout time
- DietPi-Software | Add nogui
- DietPi-Software | Simplify uninstall
- DietPi-Software | Remove G_EXEC
- DietPi-Software | Merge into one Create_Config step
- DietPi-Software | Raise timeout time again
- DietPi-Software | Fix G_CONFIG_INJECT
- DietPi-Software | Add daemon-reload to Config_Create function
- DietPi-Software | Raise timeout time a little more
- DietPi-Software | Fixes
- DietPi-Software | Move fi to correct place
- DietPi-Software | Add G_EXEC to daemon-reload
- DietPi-Software | Add RCON server for stopping/accessing
- DietPi-Software | Actually add stop command
- DietPi-Software | Fix paths
- DietPi-Software | Remove chown
- DietPi-Software | Fix more paths
- DietPi-Software | Add chmod
- DietPi-Software | Switch to numerical to match other style
- DietPi-Software | MCrcon configuration
- DietPi-Software | Fix paths again
- DietPi-Software | Fixes
- DietPi-Software | Move EULA
- DietPi-Software | Link MCrcon to /usr/local/bin for easy access
- DietPi-Software | Remove noconsole
- DietPi-Software | Re-add noconsole
- DietPi-Software | Add doc link

nonnorm committed:
Add Paper MC Server

nonnorm committed:
Add Paper MC to uninstall

MichaIng commented on dietpi/dietpi-software:
```diff
@@ -12465,6 +12484,47 @@ _EOF_
+
+[Install]
+WantedBy=multi-user.target
+_EOF_
```
Two suggestions to align with other software titles installed  via dietpi-software:
1. Install to `/opt/<name>`, e.g. `/opt/papermc`
2. Run the server as as unprivileged user, e.g. `papermc`:
```
Create_User -d /opt/papermc papermc # This creates a system user named papermc with /opt/papermc has its home dir.
```
then in systemd unit:
```
User=papermc
```

MichaIng commented on dietpi/dietpi-software:
```diff
@@ -12465,6 +12484,47 @@ _EOF_
+[Service]
+SyslogIdentifier=papermc
+WorkingDirectory=/usr/local/bin/paper
+ExecStart=/bin/bash -c 'java -jar /usr/local/bin/paper/paperclip.jar'
```
Any reason to wrap the process into another ~4 MiB bash shell? I'd just go with:
```
ExecStart=$(command -v java) -jar /path/to/paperclip.jar'
```
`'_EOF_'` => `_EOF_` required then to expand the Java executable path.

MichaIng commented on dietpi/dietpi-software:
```diff
@@ -12465,6 +12484,47 @@ _EOF_
+        if G_WHIP_YESNO "Would you like to install the Geyser and Floodgate
+        \n plugins for compatibility with Bedrock Edition?"; then
+            curl https://ci.nukkitx.com/job/GeyserMC/job/Geyser/job/master/lastSuccessfulBuild/artifact/bootstrap/spigot/target/Geyser-Spigot.jar -o /usr/local/bin/paper/plugins/Geyser-Spigot.jar
+            curl https://ci.nukkitx.com/job/GeyserMC/job/Floodgate/job/development/lastSuccessfulBuild/artifact/bukkit/target/floodgate-bukkit.jar -o /usr/local/bin/paper/plugins/floodgate-bukkit.jar
```
`Download_Install` can be used here as well to include some error handling, like for the main jar, or:
```
G_EXEC curl -sSfL <URL> </path/to/dest.jar>
```
which reduces overhead a bid (no dedicated file existence check). `f` flag btw is important to avoid curl from writing a potential 40X error handler HTML to the target file but exit with error message instead, `L` in case the download is moved to a different location in the future and a redirect is left in place.

MichaIng commented:
Awesome, many thanks for implementing this 👍. I'll do some review.

MichaIng commented on dietpi/dietpi-software:
```diff
@@ -12465,6 +12484,47 @@ _EOF_
+			G_WHIP_BUTTON_OK_TEXT='YES'
+			G_WHIP_BUTTON_CANCEL_TEXT='NO'
+			if G_WHIP_YESNO "Do you agree to the Minecraft EULA found at.
+\nhttps://account.mojang.com/documents/minecraft_eula"; then
```
We might want to ask this before even downloading the main jar during first install step? We'll merge install and config steps in the future, so no need to fulfil any strict separation between those 😉.

It could be even asked during interactive selection from dietpi-software menu, where a bunch of other questions and information is prompted based on selection. But would this would not include install via `dietpi-software install 181`, so not sufficient when we want to strictly require the agreement.

MichaIng commented:
Regarding pre-creating the config file. We could start the service the regular way and wait/watch until the config files have been created, then stop again. But it is a bid unsafe, e.g. the files might be present before anything or everything has been written to them. See how I did this for Bazarr: https://github.com/MichaIng/DietPi/blob/e32c1bb137745cac567eacb5980162ab34380fa8/dietpi/dietpi-software#L11786-L11798
On Sonarr then I needed to add a check for the last required settings that needs to be written to the config file: https://github.com/MichaIng/DietPi/blob/e32c1bb137745cac567eacb5980162ab34380fa8/dietpi/dietpi-software#L11810

So above as worst-case solution, but of course best would be to have some CLI for this which is explicitly made to pre-create configs and exits automatically.

nonnorm committed:
Fixes

MichaIng commented on dietpi/dietpi-software:
```diff
@@ -12488,41 +12495,30 @@ _EOF_
+          Download_Install 'https://ci.nukkitx.com/job/GeyserMC/job/Floodgate/job/development/lastSuccessfulBuild/artifact/bukkit/target/floodgate-bukkit.jar' /opt/papermc/plugins/floodgate-bukkit.jar
+      fi
+      # Service
+      cat << '_EOF_' > /etc/systemd/system/papermc.service
```
`'_EOF_'` needs to be changed to `_EOF_`: With single quotes around the closing string for "here documents", variables and command substitutions within the document are not expended, everything is taken literally, hence `$(command -v java)` would currently be written literally to the systemd unit instead of being expended to `/usr/bin/java` (AFAIK).

I use the single quotes whenever possible, to not have something expended accidentally but when we use a system executable where the location can be at different locations, I prefer to use `command -v <command>`, e.g. to use `/usr/local/bin/java` when Java has been manually build on the system.

nonnorm committed:
Create config files

MichaIng commented on dietpi/dietpi-software:
```diff
@@ -12488,41 +12495,30 @@ _EOF_
-SyslogIdentifier=papermc
-WorkingDirectory=/usr/local/bin/paper
-ExecStart=/bin/bash -c 'java -jar /usr/local/bin/paper/paperclip.jar'
+User=papermc
```
`SyslogIdentifier=papermc` can stay. Else in system log, messages of would be prefixed with "java" currently, before last commit it would have been "bash", hence by default it is the major executable that is called via `ExecStart`. In such cases, where one does not call the software executable directly, e.g. a `papermc` binary, for sake of beauty and overview of syslog it makes sense to define `SyslogIdentifier` to prefix the syslog entries with the name of the software service 😃.

MichaIng commented on dietpi/dietpi-software:
```diff
@@ -12519,6 +12519,22 @@ ExecStart=$(command -v java) -jar /opt/papermc/paperclip.jar'
+					sleep 1
+				done
+				sleep 1
+				G_EXEC_NOHALT=1 G_EXEC systemctl stop paper
```
Little typo here? Should be `systemctl stop papermc`, I guess 😉.

nonnorm committed:
Add floodgate authentication

nonnorm committed:
More Fixes

nonnorm committed:
Switch to G_CONFIG_INJECT

nonnorm committed:
Merge pull request #2 from MichaIng/dev

Dev

MichaIng committed:
v6.34 (#3855)

+ DietPi-Bugreport | Do not check if upload address is reachable via HTTP/HTTPS but check on SFTP/SSH port explicitly since this is how the bug report is uploaded
+ DietPi-Bugreport | Massively simplify code by using DietPi-Globals

MichaIng committed:
v6.34

+ DietPi-Imager | Fix and add menu pre-selections

nonnorm committed:
Remove apostrophe

MichaIng committed:
v6.34 (#3848)

+ DietPi-Software | OpenBazaar: Update Go to current v1.15.3
+ DietPi-Software | OpenBazaar: Run service with unprivileged user "openbazaar"
+ DietPi-Software | OpenBazaar: Disable file logging "-f" which forces logs to STDOUT (=systemd-journald); Reduce log level to "notice" to avoid verbose debug logging to system journal
+ DietPi-Software | OpenBazaar: Update it on reinstall i.e. when it has been installed already through go get
+ DietPi-Software | OpenBazaar: On uninstall, remove binary and archive via "go clean"

nonnorm committed:
Fix `fi`

nonnorm committed:
Finish fixing

nonnorm committed:
Add final fi

nonnorm commented:
I'm going to wait until we can get #3854 merged, then continue testing and debugging with the new function.

MichaIng committed:
v6.34 (#3860)

+ DietPi-Software | CloudPrint: Remove from software options
+ DietPi-Patch | CloudPrint: Do not reinstall it anymore, since this install option has been removed
+ DietPi-Patch | CloudPrint: Unset CloudPrint install flag to free up ID
+ DietPi-Survey_report | Add Bazarr with v6.33
+ DietPi-Survey_report | Add v6.34 support and remove CloudPrint with this version

MichaIng committed:
v6.34

+ DietPi-Obtain_HW_model | Fix RPi CM4 detection, the revision code ID is 14, not 12: https://github.com/raspberrypi/documentation/pull/1616

nonnorm committed:
v6.34 (#3854)

+ DietPi-Software | Add Create_Config function which starts a service and waits for a config file to be pre-created and optionally content to be present. This is useful when we need the current default config file present on a fresh install to alter it.

nonnorm committed:
Merge pull request #4 from MichaIng/dev

MichaIng commented:
Lol where the hack is this PR #4 coming from in the commit history now? 😄 However file changes are clean now 👍. Will run some tests and will merge tomorrow, too tired today to finish changelog, add it to survey report + dietpi-services etc 😉.

nonnorm commented:
This isn't actually ready, I need to use the new function.  It will be finished soon.

MichaIng commented on dietpi/dietpi-software:
```diff
@@ -1178,6 +1178,16 @@ DietPi-Software will decrypt and use it for software installs. You can change it
+    software_id=181
+    
+    aSOFTWARE_NAME[$software_id]='Paper'
+		aSOFTWARE_DESC[$software_id]='A highly optimized server for Java Edition with plugins'
```
What do you think about naming it `PaperMC` and "`Highly optimised Minecraft server with plugins, written in Java`"

Fits to the naming on their [website](https://papermc.io/) as well and at least gives a hint that it's a Minecraft server 😉.

MichaIng commented on dietpi/dietpi-software:
```diff
@@ -12376,6 +12402,63 @@ _EOF_
+      else
+        if [[ -f '/opt/papermc/plugins/Geyser-Spigot.jar' ]] && [[ -f '/opt/papermc/plugins/floodgate-bukkit.jar' ]]
+        then
+        G_CONFIG_INJECT 'auth-type: online' 'auth-type: floodgate' /opt/papermc/plugins/Geyser-Spigot/config.yml
```
Use it the following way:
```
G_CONFIG_INJECT 'auth-type:[[:blank:]]' 'auth-type: floodgate' /opt/papermc/plugins/Geyser-Spigot/config.yml
```
That way `floodgate` authentication is applied regardless which setting was present before.
`[[:blank:]]` catches any kind of space, although not sure if YAML allows anything else than regular spaces?

nonnorm committed:
Make use of new function

nonnorm committed:
Fix uninstall

nonnorm committed:
Cosmetic fixes

nonnorm committed:
Reverse CFG command

nonnorm committed:
Increase timeout

nonnorm committed:
Switch to G_CONFIG_INJECT EULA

nonnorm committed:
Add G_EXEC_OUTPUT

nonnorm committed:
Add G_EXEC_NOEXIT

nonnorm committed:
Add 1 second of sleep

nonnorm committed:
Switch back to echo

nonnorm committed:
Switch back to G_CONFIG_INJECT

nonnorm committed:
Add cd

nonnorm committed:
Give users access to /opt/papermc

nonnorm committed:
Fix uninstall again

nonnorm committed:
Increase timeout time

nonnorm committed:
Add Config_Create for Geyser CFG

nonnorm committed:
Add missing service

nonnorm committed:
Add blank line

nonnorm committed:
Add to dietpi-services

nonnorm committed:
Raise Geyser timeout time

nonnorm committed:
Add nogui

nonnorm committed:
Simplify uninstall

nonnorm committed:
Remove G_EXEC

nonnorm committed:
Merge into one Create_Config step

nonnorm committed:
Raise timeout time again

nonnorm committed:
Fix G_CONFIG_INJECT

nonnorm committed:
Add daemon-reload to Create_Config function

nonnorm committed:
Raise timeout time a little more

nonnorm committed:
Inject stop test

nonnorm committed:
Remove stop test

MichaIng commented on dietpi/dietpi-software:
```diff
@@ -12417,7 +12417,7 @@ _EOF_
       G_EXEC cd /opt/papermc
       G_EXEC_OUTPUT=1 G_EXEC java -jar /opt/papermc/paperclip.jar nogui
-      G_CONFIG_INJECT 'eula=[[:blank:]]' 'eula=true' /opt/papermc/eula.txt
+      G_CONFIG_INJECT 'eula=false' 'eula=true' /opt/papermc/eula.txt
```
Use `G_CONFIG_INJECT 'eula=' 'eula=true' /opt/papermc/eula.txt` here. `[[:blank:]]` is a placeholder/wildcard for any kind of space, but this option obviously does not contain a space.

Compare with below:
```
G_CONFIG_INJECT 'auth-type:[[:blank:]]' 'auth-type: floodgate' /opt/papermc/plugins/Geyser-Spigot/config.yml
```
Here, due to YAML syntax, a space is always present/required between the name of the setting and the value, so we can include that into the matching pattern to be sure that only the exact setting we want is matched. In this case  `'auth-type:'` would be enough as well, since a `:` in YAML is a strict separator between settings name and value, AFAIK, same as the space. However e.g. `'auth-type'` could be in theory not enough, as there might be a setting now or in the future that is named `auth-type-strict` or something like that, maybe to remove an otherwise present selector in the web interface, or that each user can/cannot choose it's own authentication type or such (just a dumb example, but you get the point).

Good practice is to define the matching pattern as specific as possible, at best oriented at the syntax of the config file type, but to not include any part of the value (true, false, floodgate) so that the setting is set, regardless which value was present before.

nonnorm committed:
Fixes

nonnorm committed:
Move fi to correct place

nonnorm committed:
Add G_EXEC to daemon-reload

nonnorm committed:
Add RCON server for stopping/accessing

nonnorm committed:
Actually add stop command

nonnorm committed:
Fix paths

nonnorm committed:
Remove chown

nonnorm committed:
Fix more paths

nonnorm committed:
Add chmod

nonnorm committed:
Switch to numerical to match other style

nonnorm committed:
Mcrcon configuration

nonnorm committed:
Fix paths again

MichaIng commented on dietpi/dietpi-software:
```diff
@@ -12376,6 +12406,56 @@ _EOF_
+      G_EXEC chown -R papermc /opt/papermc
+      G_EXEC usermod -a -G papermc root
+      G_EXEC chmod -R 760 /opt/papermc
+      Download_Install https://github.com/Tiiffi/mcrcon/archive/v0.7.1.zip /opt/papermc/ G_EXEC unzip /opt/papermc/mcrcon.zip G_EXEC rm /opt/papermc/mcrcon.zip
```
```
G_EXEC unzip /opt/papermc/mcrcon.zip G_EXEC rm /opt/papermc/mcrcon.zip
```
These four arguments have no effect and are not required. 
The `Download_Install` function only requires the URL and optionally the path to extract the archive to. unzip and archive removal are done automatically by the function 😎.

If finally only the `mcrcon` binary is required, I suggest to do:
```
Download_Install https://github.com/Tiiffi/mcrcon/archive/v0.7.1.tar.gz # gzip archive, better compressed + faster decompression, without path, so it extracts to the current tmpfs directory: /tmp/DietPi-Software
gcc -g0 -O3 mcrcon-0.7.1/mcrcon.c -o /opt/papermc/mcrcon # Keep in papermc directory? Compile options to reduce binary size and optimise execution speed
rm -R mcrcon-0.7.1
strip --remove-section=.comment --remove-section=.note /opt/papermc/mcrcon # This again reduces the file size radically
```

MichaIng commented on dietpi/dietpi-software:
```diff
@@ -12376,6 +12406,56 @@ _EOF_
+      G_CONFIG_INJECT 'eula=' 'eula=true' /opt/papermc/eula.txt
+      G_EXEC chown -R papermc /opt/papermc
+      G_EXEC usermod -a -G papermc root
+      G_EXEC chmod -R 760 /opt/papermc
```
There is no point in doing the above last two commands:
- `usermod -a -G papermc root` This adds the `root` user to the `papermc` group, but the root user has always full permissions, hence adding it to any group has no effect 😉.
- `chmod -R 760 /opt/papermc` this would then allow the root user write and read (but no execute) permissions, but even with `000` mode the root user can always do everything.

Do only: `G_EXEC chown -R papermc:papermc /opt/papermc`, so the papermc user has full permissions and, aside of root, no other user, which should be perfectly fine.

nonnorm committed:
Fixes

MichaIng commented:
Great, I'll run some tests tomorrow 👍 😃.

MichaIng commented:
#### Testing
- [x] VM Buster
- [x] VM Stretch
- [x] VM Bullseye
- [x] ARMv8

MichaIng commented on dietpi/dietpi-software:
```diff
@@ -12376,6 +12406,55 @@ _EOF_
+      G_EXEC chown -R papermc:papermc /opt/papermc
+      G_EXEC cd /opt/papermc
+      G_EXEC_OUTPUT=1 G_EXEC java -jar /opt/papermc/paperclip.jar nogui
+      G_CONFIG_INJECT 'eula=' 'eula=true' /opt/papermc/eula.txt
```
```
[ INFO ] DietPi-Software | java -jar /opt/papermc/paperclip.jar nogui, please wait...
Downloading vanilla jar...
Patching vanilla jar...
System Info: Java 11 (OpenJDK 64-Bit Server VM 11.0.9+11-post-Debian-1deb10u1) Host: Linux 4.19.0-10-amd64 (amd64)
Loading libraries, please wait...
[12:52:35 ERROR]: Failed to load properties from file: server.properties
[12:52:36 WARN]: Failed to load eula.txt
[12:52:36 INFO]: You need to agree to the EULA in order to run the server. Go to eula.txt for more info.
[  OK  ] DietPi-Software | java -jar /opt/papermc/paperclip.jar nogui
[  OK  ] DietPi-Software | Setting in /opt/papermc/eula.txt adjusted: eula=true
```
Let's do `echo 'eula=true' > /opt/papermc/eula.txt` right after EULA has been accepted in install step. The file does not contain anything else, so that is safe to do.

Not sure if `java -jar /opt/papermc/paperclip.jar nogui` is then required at all or was this only to produce the `eula.txt`?

nonnorm replied:
Probably not, it really just downloads the server JAR file.  That can be done on first run.  I'll remove it.  Also, if the EULA is pre-accepted, it will just run the server.

MichaIng replied:
That btw also solves a wrong permissions issue due to server start as root which broke at least file logging derived from syslogs:
```
2020-10-26 13:21:03 root@VM-Buster:/opt/papermc# l
total 45M
-rw-r--r-- 1 papermc papermc    2 Oct 26 12:55 banned-ips.json
-rw-r--r-- 1 papermc papermc    2 Oct 26 12:55 banned-players.json
-rw-r--r-- 1 papermc papermc 1.2K Oct 26 12:55 bukkit.yml
drwxr-xr-x 2 root    root    4.0K Oct 26 12:51 cache
-rw-r--r-- 1 papermc papermc  598 Oct 26 12:55 commands.yml
-rw-r--r-- 1 root    root     250 Oct 26 12:52 eula.txt
-rw-r--r-- 1 papermc papermc 2.6K Oct 26 12:53 help.yml
drwxr-xr-x 2 root    root    4.0K Oct 26 12:52 logs
-rwxr-xr-x 1 papermc papermc  19K Oct 26 12:51 mcrcon
-rw-r--r-- 1 papermc papermc    2 Oct 26 12:55 ops.json
-rw-r--r-- 1 papermc papermc 6.5K Oct 26 12:56 paper.yml
-rw-r--r-- 1 papermc papermc  45M Oct 26 12:51 paperclip.jar
-rw-r--r-- 1 papermc papermc    0 Oct 26 12:53 permissions.yml
drwxr-xr-x 5 papermc papermc 4.0K Oct 26 12:53 plugins
-rw-r--r-- 1 root    root    1.1K Oct 26 12:55 server.properties
-rw-r--r-- 1 papermc papermc 4.3K Oct 26 12:56 spigot.yml
-rw-r--r-- 1 papermc papermc    2 Oct 26 12:55 usercache.json
-rw-r--r-- 1 papermc papermc   47 Oct 26 12:53 version_history.json
-rw-r--r-- 1 papermc papermc    2 Oct 26 12:53 whitelist.json
drwxr-xr-x 7 papermc papermc 4.0K Oct 26 13:17 world
drwxr-xr-x 3 papermc papermc 4.0K Oct 26 13:17 world_nether
drwxr-xr-x 3 papermc papermc 4.0K Oct 26 13:17 world_the_end
```

MichaIng commented on dietpi/dietpi-software:
```diff
@@ -12440,7 +12440,7 @@ _EOF_
       if [[ -f '/opt/papermc/plugins/Geyser-Spigot.jar' ]] && [[ -f '/opt/papermc/plugins/floodgate-bukkit.jar' ]]
       then
-        Create_Config /opt/papermc/plugins/Geyser-Spigot/config.yml papermc 110
+        Create_Config /opt/papermc/plugins/Geyser-Spigot/config.yml papermc 115
```
Where did you tested it to get this long required time? On my VM it took something around 10 seconds. I just thought it would be nice to report the time it took to create the config. For end user probably not important but for developers, when it's about tuning the timeouts, its good to know and review how long it took on the different machines 🙂.

nonnorm replied:
It took a minimum of about 104 seconds on my VM (VirtualBox), with the only modification being the extra GiB of RAM.  A maximum of about 111/112.

MichaIng replied:
Wow that is long, okay not much harm to allow a longer timeout. This is an interactive-only install so users transparently see what's happening and the counter. Also with the above server start now removed, this one will likely take longer again 😉.

nonnorm committed:
Move EULA

MichaIng commented:
Wow 815 MiB RAM usage, that's Java... I'm always wondering when it states "highly optimized" + Java which is not going well into my head. I mean there are quite some Java applications which seem to load only the very required subset of Java classes and hence use much less memory, but in every other case, no matter how tiny the actual program is, up to 1 GiB is usual, insane...

It does not say something about the performance of course and especially the scalability, as probably active usage does not raise memory consumption that much anymore, but at least no change to reasonably run it on <= 1 GiB SBCs 😄. I'm never sure if simply using something like `-Xmx100M` works fine, so that a high default is only reserved but not actually used?
**EDIT: https://aikar.co/2018/07/02/tuning-the-jvm-g1gc-garbage-collector-flags-for-minecraft/**
> Use these flags exactly, only changing Xmx and Xms. These flags work and scale accordingly to any size of memory, even 500MB but 1.15 will not do well with such low memory…)

CPU usage constantly around 20% on one core.
__________
But otherwise it seems to work fine. I am not experienced with Minecraft at all @ravenclaw900 are there any reasonable status commands, e.g. via `mcrcon`?
__________
```
Oct 26 13:26:10 VM-Buster papermc[5903]: [13:26:10 WARN]: To start the server with more ram, launch it as "java -Xmx1024M -Xms1024M -jar minecraft_server.jar"
...
Oct 26 13:26:27 VM-Buster papermc[5903]: [13:26:27 WARN]: [Geyser-Spigot] en_ is not a valid Bedrock language.
...
Oct 26 13:26:28 VM-Buster papermc[5903]: [13:26:28 WARN]: [org.geysermc.platform.spigot.shaded.reflections.Reflections] given scan urls are empty. set urls in the configuration
Oct 26 13:26:33 VM-Buster papermc[5903]: [13:26:33 WARN]: [org.geysermc.platform.spigot.shaded.reflections.Reflections] given scan urls are empty. set urls in the configuration
Oct 26 13:27:07 VM-Buster papermc[5903]: [13:27:07 WARN]: [org.geysermc.platform.spigot.shaded.reflections.Reflections] given scan urls are empty. set urls in the configuration
Oct 26 13:27:08 VM-Buster papermc[5903]: [13:27:08 WARN]: [org.geysermc.platform.spigot.shaded.reflections.Reflections] given scan urls are empty. set urls in the configuration
Oct 26 13:27:10 VM-Buster papermc[5903]: [13:27:10 WARN]: [org.geysermc.platform.spigot.shaded.reflections.Reflections] given scan urls are empty. set urls in the configuration
```
- I cannot say how much memory is required to reasonably run this, or which are actually the defaults (read 64 MiB somewhere but that is obviously wrong). Probably that is something we leave to end user, as it highly depends on usage and machine anyway.
- Any idea where to set the language string?
- Not sure what the scan URLs are about and if it would be reasonable to pre-configure this?
____________
And another idea. Since I didn't find a way to disable or change file logging locations, we could move it either to `/var/log/papermc` by via symlink or try to disable it via symlink to `/dev/null`, as it logs to syslog (`jourrnalctl -u papermc`) anyway. But it shall not produce error messages as this irritates users.

nonnorm commented:
No inbuilt status commands, probably a plugin that you can install though.  About the scan URLs, see https://github.com/GeyserMC/Geyser/issues/1252.

MichaIng commented on dietpi/dietpi-software:
```diff
@@ -6343,6 +6357,23 @@ exec sudo -u $ha_user dash -c '$ha_pyenv_activation; exec pip3 install -U homeas
+			if G_WHIP_YESNO "Do you agree to the Minecraft EULA found at:
+\nhttps://account.mojang.com/documents/minecraft_eula"; then
+        Download_Install 'https://papermc.io/ci/job/Paper-1.16/lastSuccessfulBuild/artifact/paperclip.jar' /opt/papermc/paperclip.jar
+        G_EXEC echo 'eula=true' > /opt/papermc/eula.txt
```
This does not work, `G_EXEC` cannot be used with redirects. Run this command from console and check `/opt/papermc/eula.txt` content 😉. Workaround:
```
G_EXEC eval 'echo "eula=true" > /opt/papermc/eula.txt'
```
Or `echo 'eula=true' > /opt/papermc/eula.txt` only without error handler.

I also didn't find a final good solution for this, probably with G_EXEC_REDIRECT or something as there are general concerns of using `eval`, especially in combination with variables.

nonnorm replied:
I noticed that, that’s why I switched to G_CONFIG_INJECT.  I’ll use veal.

MichaIng replied:
Done already, yes `G_CONFIG_INJECT` would have been an alternative but the file needs to exist, at least empty.

MichaIng committed:
v6.34

+ DietPi-Services | Fix indentation

MichaIng committed:
v6.34

+ DietPi-Software | PaperMC: Fix indentation and complement uninstall

MichaIng committed:
v6.34

+ DietPi-Software | Syntax

MichaIng committed:
v6.34

+ DietPi-Software | Create_Config(): Adjust default timeout to mostly used 25 seconds and print how long it actually took

nonnorm commented:
Sorry about the indentation, it all seems to be correct on my computer.

MichaIng commented:
No problem, it is a question of the default editor settings, which can be even changed in GitHub web UI text editor. Either tabs are added as real tab characters or as a configurable amount of spaces (in your case it was two spaces per tab). And then the tab character can be shown with a configurable width, in your editor obviously with a width of two unicode characters while on GitHub 8 ones. A bid of a pain as when we go with spaces, contributors might add real tabs. The argument for real tabs is that it can be configured with how much width they are shown, hence just according to individual preferences.

nonnorm commented:
For the language thing, I think that might because the default C.UTF-8 locale is being used.  I use en_US, and I get:
`[Geyser-Spigot] Downloading Minecraft JAR to extract en_us locale, please wait... (this may take some time depending on the speed of your internet connection)`

MichaIng commented:
I found the setting `default-locale` in `plugins/Geyser-Spigot/config.yml` and setting it to `en_gb` mutes the warning as well. Interestingly nothing is downloaded in my case, at least no message about this.

However makes sense then to keep it commended so that either the configured console locale is used automatically or users can override when studying the available config files.

Okay I have two more suggestions:
1. Changing the users home directory and systemd service `WorkingDirectory` to `/mnt/dietpi_userdata/papermc`. This separates the raw server jar from config/data:
    ```
    2020-10-26 15:55:42 root@VM-Stretch:/mnt/dietpi_userdata/papermc# l
    total 84K
    -rw-r--r-- 1 papermc papermc    2 Oct 26 15:54 banned-ips.json
    -rw-r--r-- 1 papermc papermc    2 Oct 26 15:54 banned-players.json
    -rw-r--r-- 1 papermc papermc 1.2K Oct 26 15:54 bukkit.yml
    drwxr-xr-x 2 papermc papermc 4.0K Oct 26 15:40 cache
    -rw-r--r-- 1 papermc papermc  598 Oct 26 15:54 commands.yml
    -rw-r--r-- 1 papermc papermc   10 Oct 26 15:42 eula.txt
    -rw-r--r-- 1 papermc papermc 2.6K Oct 26 15:42 help.yml
    drwxr-xr-x 2 papermc papermc 4.0K Oct 26 15:53 logs
    -rw-r--r-- 1 papermc papermc    2 Oct 26 15:54 ops.json
    -rw-r--r-- 1 papermc papermc 6.5K Oct 26 15:54 paper.yml
    -rw-r--r-- 1 papermc papermc    0 Oct 26 15:42 permissions.yml
    drwxr-xr-x 5 papermc papermc 4.0K Oct 26 14:55 plugins
    -rw-r--r-- 1 papermc papermc 1.1K Oct 26 15:53 server.properties
    -rw-r--r-- 1 papermc papermc 4.3K Oct 26 15:54 spigot.yml
    -rw-r--r-- 1 papermc papermc    2 Oct 26 15:54 usercache.json
    -rw-r--r-- 1 papermc papermc   47 Oct 26 15:42 version_history.json
    -rw-r--r-- 1 papermc papermc    2 Oct 26 15:42 whitelist.json
    drwxr-xr-x 7 papermc papermc 4.0K Oct 26 16:00 world
    drwxr-xr-x 3 papermc papermc 4.0K Oct 26 16:00 world_nether
    drwxr-xr-x 3 papermc papermc 4.0K Oct 26 16:00 world_the_end
    2020-10-26 16:00:48 root@VM-Stretch:/mnt/dietpi_userdata/papermc# l /opt/papermc/
    total 45M
    -rwxr-xr-x 1 papermc papermc 19K Oct 26 14:53 mcrcon
    -rw-r--r-- 1 papermc papermc 45M Oct 26 14:53 paperclip.jar
    ```
    `/opt/papermc` could stay root:root owner:group then. Benefit is mostly for SBCs with system on SD card to reduce SD card writes and speed up I/O in general when having `/mnt/dietpi_userdata` moved to an external USB drive. It matches how we setup most software titles.
2. Since I didn't find a way to disable file logging, linking the `logs` dir to `/var/log/papermc` to have them in RAMlog, if configured, else in combined log directory at least.

What do you think?

nonnorm commented:
Those all sound good to me.

MichaIng committed:
v6.34

+ DietPi-Software | PaperMC: Move config and data to /mnt/dietpi_userdata/papermc
+ DietPi-Software | PaperMC: Link logs to /var/log/papermc
+ DietPi-Software | PaperMC: Use "lastStable" builds from Jenkins instead of "lastSuccessful" and latest release of mcrcon from master branch
+ DietPi-Software | Minor code/format enhancements

MichaIng commented:
```
[ INFO ] DietPi-Software | Papermc config file got created after 117 seconds
```
Let's see, I think for ARMs we need to raise the timeout 😄.

One issue: When the service is stopped after config file creation, the mcrcon password has not yet been set, hence stopping fails, respectively it times out and SIGKILL does the job uncleanly.

I'm not sure how to solve this best. We could alter the service file and add ExecStop after config file creation, so it is killed a bid unclean but at least via SIGTERM (which allows the process to shutdown itself) instead of SIGKILL after 90 seconds timeout, which is a hard kill, most unclean.

MichaIng committed:
v6.34

+ DietPi-Software | PaperMC: Pre-create log directory

nonnorm commented:
I found that, in my testing, SIGTERM doesn’t work.  It delays the 1:30 timeout, then goes to SIGKILL.

MichaIng commented:
Hmm, and there is also no command line option or anything similar. So the question is how can be cleanly shutdown the service without mcrcon then 🤔?
Pre-creating the config file (with only those two settings) most likely leads to it being otherwise empty, so no defaults/settings exposed afterwards.

I'll play a bid around. It's not beautiful but not something that prevents us from implementing it like that. It's then basically something we should report to the developers. A SIGTERM IMO must be handled successfully, that is the basis of a clean exit of any software where data is stored on disk.

MichaIng committed:
v6.34

+ DietPi-Software | PaperMC: Mask rcon password in console output

nonnorm commented:
It might not be Paper’s fault.  Probably Java’s.

MichaIng commented:
Doesn't make much sense to talk about "fault" here at all, this is not what I wanted to say 😉. However, I know that Java in a way forwards signals (SIGTERM) since other Java applications react gracefully on those, so it's on PaperMC devs to have a look how to catch and handle it, if a clean exit from outside of an interactive console on UNIX is wanted, even that it might be made difficult/non-intuitive by Java.

Java devs/package developers on the other hand need to have a look into Java on armhf, probably in containers, as it fails with segmentation fault when trying it on RPi/Raspbian in systemd-nspawn container 🤔. On x86 VM aarch64 systemd-nspawn container it works 🤔. However, I'm sure that it's only an issue in combination with the container.

nonnorm commented:
Also not really what I meant to say...  Anyway, apparently Paper already should?  https://github.com/PaperMC/Paper/issues/4639#issuecomment-706560894

MichaIng commented:
Okay it seems to generally work, but it fails when the server has not fully start up yet:
```
Oct 26 21:05:34 VM-Buster papermc[6547]: [21:05:34 INFO]: Preparing start region for dimension minecraft:overworld
Oct 26 21:05:38 VM-Buster papermc[6547]: [21:05:38 INFO]: Loaded 0 spawn chunks for world world
Oct 26 21:05:38 VM-Buster papermc[6547]: [21:05:38 INFO]: Preparing spawn area: 0%
Oct 26 21:05:38 VM-Buster papermc[6547]: [21:05:38 INFO]: Preparing spawn area: 0%
Oct 26 21:05:39 VM-Buster systemd[1]: Stopping papermc (DietPi)...
Oct 26 21:05:39 VM-Buster papermc[6547]: [21:05:39 INFO]: Preparing spawn area: 0%
Oct 26 21:05:39 VM-Buster papermc[6547]: [21:05:39 INFO]: Preparing spawn area: 0%
Oct 26 21:05:40 VM-Buster papermc[6547]: [21:05:40 INFO]: Preparing spawn area: 0%
Oct 26 21:05:40 VM-Buster papermc[6547]: [21:05:40 INFO]: Stopping main thread (Ignore any thread death message you see! - DO NOT REPORT THREAD DEATH TO PAPER)
Oct 26 21:05:40 VM-Buster papermc[6547]: [21:05:40 INFO]: Preparing spawn area: 0%
Oct 26 21:05:40 VM-Buster papermc[6547]: [21:05:40 ERROR]: Main thread terminated by WatchDog due to hard crash
Oct 26 21:05:40 VM-Buster papermc[6547]: java.lang.ThreadDeath: null
Oct 26 21:05:40 VM-Buster papermc[6547]:         at java.lang.Thread.stop(Thread.java:942) [?:?]
Oct 26 21:05:40 VM-Buster papermc[6547]:         at net.minecraft.server.v1_16_R2.MinecraftServer.stop(MinecraftServer.java:761) ~[patched_1.16.3.jar:git-Paper-247]
Oct 26 21:05:40 VM-Buster papermc[6547]:         at net.minecraft.server.v1_16_R2.DedicatedServer.stop(DedicatedServer.java:698) ~[patched_1.16.3.jar:git-Paper-247]
Oct 26 21:05:40 VM-Buster papermc[6547]:         at net.minecraft.server.v1_16_R2.MinecraftServer.close(MinecraftServer.java:735) ~[patched_1.16.3.jar:git-Paper-247]
Oct 26 21:05:40 VM-Buster papermc[6547]:         at org.bukkit.craftbukkit.v1_16_R2.util.ServerShutdownThread.run(ServerShutdownThread.java:28) ~[patched_1.16.3.jar:git-Paper-247]
Oct 26 21:05:40 VM-Buster papermc[6547]: [21:05:40 INFO]: Stopping server
Oct 26 21:05:40 VM-Buster papermc[6547]: [21:05:40 INFO]: Saving players
Oct 26 21:05:40 VM-Buster papermc[6547]: [21:05:40 INFO]: Saving worlds
Oct 26 21:05:40 VM-Buster papermc[6547]: [21:05:40 INFO]: Saving chunks for level 'ServerLevel[world]'/minecraft:overworld
Oct 26 21:05:43 VM-Buster papermc[6547]: [21:05:43 INFO]: Preparing spawn area: 0%
Oct 26 21:05:43 VM-Buster papermc[6547]: [21:05:43 INFO]: Preparing spawn area: 0%
Oct 26 21:05:43 VM-Buster papermc[6547]: [21:05:43 INFO]: Preparing spawn area: 0%
Oct 26 21:05:43 VM-Buster papermc[6547]: [21:05:43 INFO]: Preparing spawn area: 0%
Oct 26 21:05:43 VM-Buster papermc[6547]: [21:05:43 INFO]: Preparing spawn area: 0%
Oct 26 21:05:43 VM-Buster papermc[6547]: [21:05:43 INFO]: Preparing spawn area: 0%
Oct 26 21:05:44 VM-Buster papermc[6547]: [21:05:44 INFO]: Preparing spawn area: 34%
Oct 26 21:05:44 VM-Buster papermc[6547]: [21:05:44 INFO]: Preparing spawn area: 51%
Oct 26 21:05:45 VM-Buster papermc[6547]: [21:05:45 INFO]: Preparing spawn area: 78%
Oct 26 21:05:45 VM-Buster papermc[6547]: [21:05:45 INFO]: Preparing spawn area: 85%
Oct 26 21:05:46 VM-Buster papermc[6547]: [21:05:46 INFO]: Preparing spawn area: 90%
Oct 26 21:05:46 VM-Buster papermc[6547]: [21:05:46 INFO]: Preparing spawn area: 90%
Oct 26 21:06:03 VM-Buster papermc[6547]: [21:06:03 INFO]: Preparing spawn area: 98%
Oct 26 21:06:03 VM-Buster papermc[6547]: [21:06:03 INFO]: Preparing spawn area: 98%
Oct 26 21:06:03 VM-Buster papermc[6547]: [21:06:03 INFO]: Preparing spawn area: 98%
Oct 26 21:06:03 VM-Buster papermc[6547]: [21:06:03 INFO]: Preparing spawn area: 98%
Oct 26 21:06:03 VM-Buster papermc[6547]: [21:06:03 INFO]: Preparing spawn area: 98%
Oct 26 21:06:03 VM-Buster papermc[6547]: [21:06:03 INFO]: Preparing spawn area: 98%
Oct 26 21:06:03 VM-Buster papermc[6547]: [21:06:03 INFO]: Preparing spawn area: 98%
Oct 26 21:06:03 VM-Buster papermc[6547]: [21:06:03 INFO]: Preparing spawn area: 98%
Oct 26 21:06:03 VM-Buster papermc[6547]: [21:06:03 INFO]: Preparing spawn area: 98%
Oct 26 21:06:03 VM-Buster papermc[6547]: [21:06:03 INFO]: Preparing spawn area: 98%
Oct 26 21:06:03 VM-Buster papermc[6547]: [21:06:03 INFO]: Preparing spawn area: 98%
Oct 26 21:06:03 VM-Buster papermc[6547]: [21:06:03 INFO]: Preparing spawn area: 98%
Oct 26 21:06:03 VM-Buster papermc[6547]: [21:06:03 INFO]: Preparing spawn area: 98%
Oct 26 21:06:03 VM-Buster papermc[6547]: [21:06:03 INFO]: Preparing spawn area: 98%
Oct 26 21:06:03 VM-Buster papermc[6547]: [21:06:03 INFO]: Preparing spawn area: 98%
Oct 26 21:06:03 VM-Buster papermc[6547]: [21:06:03 INFO]: Preparing spawn area: 98%
Oct 26 21:06:03 VM-Buster papermc[6547]: [21:06:03 INFO]: Preparing spawn area: 98%
Oct 26 21:06:03 VM-Buster papermc[6547]: [21:06:03 INFO]: Preparing spawn area: 98%
Oct 26 21:06:03 VM-Buster papermc[6547]: [21:06:03 INFO]: Preparing spawn area: 98%
Oct 26 21:06:03 VM-Buster papermc[6547]: [21:06:03 INFO]: Preparing spawn area: 98%
Oct 26 21:06:03 VM-Buster papermc[6547]: [21:06:03 INFO]: Preparing spawn area: 98%
Oct 26 21:06:03 VM-Buster papermc[6547]: [21:06:03 INFO]: Preparing spawn area: 98%
Oct 26 21:06:03 VM-Buster papermc[6547]: [21:06:03 INFO]: Preparing spawn area: 98%
Oct 26 21:06:03 VM-Buster papermc[6547]: [21:06:03 INFO]: Preparing spawn area: 98%
Oct 26 21:06:03 VM-Buster papermc[6547]: [21:06:03 INFO]: Preparing spawn area: 98%
Oct 26 21:06:04 VM-Buster papermc[6547]: [21:06:04 INFO]: Preparing spawn area: 98%
Oct 26 21:06:04 VM-Buster papermc[6547]: [21:06:04 INFO]: Preparing spawn area: 98%
Oct 26 21:06:05 VM-Buster papermc[6547]: [21:06:05 INFO]: Preparing spawn area: 98%
Oct 26 21:06:05 VM-Buster papermc[6547]: [21:06:05 INFO]: Preparing spawn area: 98%
Oct 26 21:06:06 VM-Buster papermc[6547]: [21:06:06 INFO]: Preparing spawn area: 98%
Oct 26 21:06:06 VM-Buster papermc[6547]: [21:06:06 INFO]: Preparing spawn area: 98%
Oct 26 21:06:07 VM-Buster papermc[6547]: [21:06:07 INFO]: Preparing spawn area: 98%
Oct 26 21:06:07 VM-Buster papermc[6547]: [21:06:07 INFO]: Preparing spawn area: 98%
Oct 26 21:06:08 VM-Buster papermc[6547]: [21:06:08 INFO]: Preparing spawn area: 98%
Oct 26 21:06:08 VM-Buster papermc[6547]: [21:06:08 INFO]: Preparing spawn area: 98%
Oct 26 21:06:09 VM-Buster papermc[6547]: [21:06:09 INFO]: Preparing spawn area: 98%
Oct 26 21:06:09 VM-Buster papermc[6547]: [21:06:09 INFO]: Preparing spawn area: 98%
Oct 26 21:06:15 VM-Buster papermc[6547]: [21:06:15 INFO]: Preparing spawn area: 98%
Oct 26 21:06:15 VM-Buster papermc[6547]: [21:06:15 INFO]: Preparing spawn area: 98%
Oct 26 21:06:15 VM-Buster papermc[6547]: [21:06:15 INFO]: Preparing spawn area: 98%
Oct 26 21:06:15 VM-Buster papermc[6547]: [21:06:15 INFO]: Preparing spawn area: 98%
Oct 26 21:06:15 VM-Buster papermc[6547]: [21:06:15 INFO]: Preparing spawn area: 98%
Oct 26 21:06:15 VM-Buster papermc[6547]: [21:06:15 INFO]: Preparing spawn area: 98%
Oct 26 21:06:15 VM-Buster papermc[6547]: [21:06:15 INFO]: Preparing spawn area: 98%
Oct 26 21:06:15 VM-Buster papermc[6547]: [21:06:15 INFO]: Preparing spawn area: 98%
Oct 26 21:06:15 VM-Buster papermc[6547]: [21:06:15 INFO]: Preparing spawn area: 98%
Oct 26 21:06:15 VM-Buster papermc[6547]: [21:06:15 INFO]: Preparing spawn area: 98%
Oct 26 21:06:15 VM-Buster papermc[6547]: [21:06:15 INFO]: Preparing spawn area: 98%
Oct 26 21:06:17 VM-Buster papermc[6547]: [21:06:17 INFO]: Preparing spawn area: 98%
Oct 26 21:06:18 VM-Buster papermc[6547]: [21:06:18 INFO]: Preparing spawn area: 98%
Oct 26 21:06:18 VM-Buster papermc[6547]: [21:06:18 INFO]: Preparing spawn area: 98%
Oct 26 21:06:19 VM-Buster papermc[6547]: [21:06:19 INFO]: Preparing spawn area: 98%
Oct 26 21:06:19 VM-Buster papermc[6547]: [21:06:19 INFO]: Preparing spawn area: 98%
Oct 26 21:06:20 VM-Buster papermc[6547]: [21:06:20 INFO]: Preparing spawn area: 98%
Oct 26 21:06:20 VM-Buster papermc[6547]: [21:06:20 INFO]: Preparing spawn area: 99%
Oct 26 21:06:21 VM-Buster papermc[6547]: [21:06:21 INFO]: Preparing spawn area: 99%
Oct 26 21:06:21 VM-Buster papermc[6547]: [21:06:21 INFO]: Preparing spawn area: 99%
Oct 26 21:06:22 VM-Buster papermc[6547]: [21:06:22 INFO]: Preparing spawn area: 99%
Oct 26 21:06:22 VM-Buster papermc[6547]: [21:06:22 INFO]: Preparing spawn area: 99%
Oct 26 21:06:23 VM-Buster papermc[6547]: [21:06:23 INFO]: Preparing spawn area: 99%
Oct 26 21:06:23 VM-Buster papermc[6547]: [21:06:23 INFO]: Preparing spawn area: 99%
Oct 26 21:06:24 VM-Buster papermc[6547]: [21:06:24 INFO]: Preparing spawn area: 100%
Oct 26 21:06:24 VM-Buster papermc[6547]: [21:06:24 INFO]: Preparing spawn area: 100%
Oct 26 21:06:25 VM-Buster papermc[6547]: [21:06:25 INFO]: Preparing spawn area: 100%
Oct 26 21:06:25 VM-Buster papermc[6547]: [21:06:25 INFO]: Preparing spawn area: 100%
Oct 26 21:06:26 VM-Buster papermc[6547]: [21:06:26 INFO]: Preparing spawn area: 100%
Oct 26 21:06:26 VM-Buster papermc[6547]: [21:06:26 INFO]: Preparing spawn area: 100%
Oct 26 21:06:27 VM-Buster papermc[6547]: [21:06:27 INFO]: Preparing spawn area: 100%
Oct 26 21:06:27 VM-Buster papermc[6547]: [21:06:27 INFO]: Preparing spawn area: 100%
Oct 26 21:06:28 VM-Buster papermc[6547]: [21:06:28 INFO]: Preparing spawn area: 100%
Oct 26 21:06:28 VM-Buster papermc[6547]: [21:06:28 INFO]: Preparing spawn area: 100%
Oct 26 21:06:29 VM-Buster papermc[6547]: [21:06:29 INFO]: Preparing spawn area: 100%
Oct 26 21:06:29 VM-Buster papermc[6547]: [21:06:29 INFO]: Preparing spawn area: 100%
Oct 26 21:06:30 VM-Buster papermc[6547]: [21:06:30 INFO]: Preparing spawn area: 100%
Oct 26 21:06:30 VM-Buster papermc[6547]: [21:06:30 INFO]: Preparing spawn area: 100%
Oct 26 21:06:31 VM-Buster papermc[6547]: [21:06:31 INFO]: Preparing spawn area: 100%
Oct 26 21:06:31 VM-Buster papermc[6547]: [21:06:31 INFO]: Preparing spawn area: 100%
Oct 26 21:06:32 VM-Buster papermc[6547]: [21:06:32 INFO]: Preparing spawn area: 100%
Oct 26 21:07:09 VM-Buster systemd[1]: papermc.service: State 'stop-sigterm' timed out. Killing.
Oct 26 21:07:09 VM-Buster systemd[1]: papermc.service: Killing process 6547 (java) with signal SIGKILL.
Oct 26 21:07:09 VM-Buster systemd[1]: papermc.service: Main process exited, code=killed, status=9/KILL
Oct 26 21:07:09 VM-Buster systemd[1]: papermc.service: Failed with result 'timeout'.
Oct 26 21:07:09 VM-Buster systemd[1]: Stopped papermc (DietPi).
```
The SIGTERM comes during spawn area preparation and that causes an error and then seems to hang or take much longer than usual 🤔.

But here the successful example:
```
Oct 26 20:21:18 VM-Buster papermc[6347]: [20:21:18 INFO]: Done (156.776s)! For help, type "help"
Oct 26 20:21:18 VM-Buster papermc[6347]: [20:21:18 INFO]: Timings Reset
Oct 26 20:31:23 VM-Buster papermc[6347]: [20:31:23 WARN]: Can't keep up! Is the server overloaded? Running 5056ms or 101 ticks behind
Oct 26 21:03:48 VM-Buster systemd[1]: Stopping papermc (DietPi)...
Oct 26 21:03:48 VM-Buster papermc[6347]: [21:03:48 INFO]: Stopping server
Oct 26 21:03:48 VM-Buster papermc[6347]: [21:03:48 INFO]: Saving players
Oct 26 21:03:49 VM-Buster papermc[6347]: [21:03:49 INFO]: Saving worlds
Oct 26 21:03:49 VM-Buster papermc[6347]: [21:03:49 INFO]: Saving chunks for level 'ServerLevel[world]'/minecraft:overworld
Oct 26 21:03:49 VM-Buster papermc[6347]: [21:03:49 INFO]: ThreadedAnvilChunkStorage (world): All chunks are saved
Oct 26 21:03:49 VM-Buster papermc[6347]: [21:03:49 INFO]: Saving chunks for level 'ServerLevel[world_nether]'/minecraft:the_nether
Oct 26 21:03:49 VM-Buster papermc[6347]: [21:03:49 INFO]: ThreadedAnvilChunkStorage (DIM-1): All chunks are saved
Oct 26 21:03:49 VM-Buster papermc[6347]: [21:03:49 INFO]: Saving chunks for level 'ServerLevel[world_the_end]'/minecraft:the_end
Oct 26 21:03:49 VM-Buster papermc[6347]: [21:03:49 INFO]: ThreadedAnvilChunkStorage (DIM1): All chunks are saved
Oct 26 21:03:49 VM-Buster papermc[6347]: [21:03:49 INFO]: ThreadedAnvilChunkStorage (world): All chunks are saved
Oct 26 21:03:49 VM-Buster papermc[6347]: [21:03:49 INFO]: ThreadedAnvilChunkStorage (DIM-1): All chunks are saved
Oct 26 21:03:49 VM-Buster papermc[6347]: [21:03:49 INFO]: ThreadedAnvilChunkStorage (DIM1): All chunks are saved
Oct 26 21:03:50 VM-Buster papermc[6347]: [21:03:50 INFO]: Flushing Chunk IO
Oct 26 21:03:50 VM-Buster papermc[6347]: [21:03:50 INFO]: Closing Thread Pool
Oct 26 21:03:50 VM-Buster papermc[6347]: [21:03:50 INFO]: Closing Server
```
______________
`mcrcon` hangs just the same during `Preparing spawn area:` phase. Looks like this is something we should report, so that this phase is cancelled successfully.

Was `mcrcon` only for stopping the service gracefully or do you think we should keep it in every case to send commands to the server? Not sure how common/useful this is?
______________
Btw, I reduced the VM power:
```
[20:21:18 INFO]: Done (156.776s)! For help, type "help"
```
We need to allow more time, also the pre-configuration hence failed 😉.

MichaIng committed:
v6.34

+ DietPi-Software | PaperMC: Do not use mcrcon to stop the server as SIGTERM has exactly the same effect
+ DietPi-Software | PaperMC: Do not initialise console as it can anyway never be used through systemd anyway
+ DietPi-Software | PaperMC: Allow more time pre-create spigot config file and do not attempt to edit it if it timed out

nonnorm commented:
MCrcon was mainly for shutting down gracefully, however, it would be very useful to have access to the server's console, at the very least to make yourself an operator.

nonnorm committed:
Link MCrcon to /usr/local/bin for easy access

nonnorm committed:
Remove noconsole

MichaIng commented:
Ah breaks "--noconsole" as well mcrcon?

nonnorm commented:
Not actually sure.  Let me test it.

MichaIng commented:
I added the flag as STDIN is anyway inaccessible through systemd and it does slightly reduce overhead.

nonnorm committed:
Re-add noconsole

nonnorm commented:
It actually seems like it doesn't make a difference to MCrcon.  I've added it back in.

MichaIng commented:
Okay let me prepare docs:
- Default server/query port: `25565`
- Install dir: `/opt/papermc`
- Config/data dir: `/mnt/dietpi_userdata/papermc`
- Main config file: `/mnt/dietpi_userdata/papermc/server.properties`
- Official docs: https://paper.readthedocs.io/
- Official sources: https://github.com/PaperMC/Paper
- Logs:
  - `journalctl -u papermc`
  - `/var/log/papermc/`

nonnorm commented:
If you are using Geyser, port 19132 is also taken.
RCON takes port 25575.

MichaIng commented:
Here it is: https://dietpi.com/phpbb/viewtopic.php?p=28177#p28177
Rough but a start. Any suggestions to extend?
We need to add it to the new docs page as well: https://github.com/MichaIng/DietPi-Docs/blob/master/docs/software/gaming.md

nonnorm commented:
Maybe add where to find plugins? (https://www.spigotmc.org/resources/)  Also, the port taken when using Geyser.

nonnorm committed:
Add doc link

MichaIng committed:
v6.34

+ DietPi-Software | Create_Config: Minor output enhancement
+ DietPi-Software | PaperMC: Install mcrcon to /usr/local/bin/ directly, in theory it can be used with any Minecraft server (?) and is not used by the PaperMC service itself anymore.

MichaIng commented:
Added 👍.

MichaIng commented:
Okay I'll merge, we can fine tune later.
